pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,300 @@
+// similar attributes and is identified by a pair of strings: the name of the FlowSchema and a "flow distinguisher".
+type FlowSchema struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+	// Specification of the desired behavior of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec FlowSchemaSpec
+	// Current status of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status FlowSchemaStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FlowSchemaList is a list of FlowSchema objects.
+type FlowSchemaList struct {
+	metav1.TypeMeta
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+
+	// Items is a list of flow-schemas.
+	Items []FlowSchema
+}
+
+// FlowSchemaSpec describes how the flow-schema's specification looks like.
+type FlowSchemaSpec struct {
+	// PriorityLevelConfigurationRef can reference a PriorityLevelConfiguration in the cluster. If the reference cannot
+	// be resolved, the flow-schema will ignored and marked as invalid in its status.
+	PriorityLevelConfigurationRef PriorityLevelConfigurationRef
+	// MatchingPriority controls the ordering the flow-schemas in the cluster. Its value must be greater than zero. The
+	// primary sort is based on MatchingPriority, ordered highest to lowest (20 before 10).
+	MatchingPriority int32
+	// Distinguisher defines how to classify requests into certain flows. If not empty, the flow's identifier is computed
+	// according to specification of the distinguisher.
+	// +optional
+	Distinguisher *FlowDistinguisher
+	// Rules holds all the RuleWithUser for this FlowSchema.
+	Rules []RuleWithUser
+}
+
+// FlowDistinguisher specifies the configuration of the distinguisher.
+type FlowDistinguisher struct {
+	// Source defines the source attribute we adopt to compute the identifier for the flow.
+	Source FlowDistinguisherSource
+}
+
+// FlowDistinguisherSource defines the input source of distinguishing flows.
+type FlowDistinguisherSource string
+
+// PriorityLevelConfigurationRef contains information that points to the request-priority being used.
+type PriorityLevelConfigurationRef struct {
+	// Name is the name of resource being referenced
+	Name string
+}
+
+// Rule is a trio of User, Verb and RequestScope.
+type RuleWithUser struct {
+	// User is is the user that this rule cares about. A nil user means the rule will be applied to all users.
+	// +optional
+	User *UserInfo
+	// Verbs is a list of verbs the rule applies to. An empty set means that everything is allowed.
```

Comment:
"empty set means that everything is allowed".  That is irregular.  The natural meaning of empty set would be "nothing is allowed".

Reply:
> "empty set means that everything is allowed". That is irregular. The natural meaning of empty set would be "nothing is allowed".

good point. i adjusted api model, now the empty fields means "nothing"

Reply:
Huh?  I still see `An empty set means that everything is allowed`.  The usual way to solve this is to prescribe a special string that means "all verbs are allowed".  In fact, it looks like such a value is already defined, with the name `VerbAll` --- you only need to mention here that it can be used here.